File: Level 3/가장 먼 노드/solution.js
function solution(n, edge) {
  const graph = Array.from({ length: n + 1 }, () => []);

  edge.forEach(([a, b]) => {
    graph[a].push(b);
    graph[b].push(a);
  });

  const queue = [[1, 0]];
  const distances = Array(n + 1).fill(0);
  const visited = Array(n + 1).fill(false);
  visited[1] = true;

  while (queue.length) {
    const [node, distance] = queue.shift();
    distances[node] = distance;

    graph[node].forEach((nextNode) => {
      if (!visited[nextNode]) {
        queue.push([nextNode, distance + 1]);
        visited[nextNode] = true;
      }
    });
  }

  const maxDistance = Math.max(...distances);

  return distances.filter((distance) => distance === maxDistance).length;
}

File: Level 3/가장 먼 노드/solution.py
from collections import deque


def build_graph(n, edge):
    graph = [[] for _ in range(n + 1)]

    for a, b in edge:
        graph[a].append(b)
        graph[b].append(a)

    return graph


def solution(n, edge):
    graph = build_graph(n, edge)

    queue = deque([(1, 0)])
    distances = [0 for _ in range(n + 1)]
    visited = [False for _ in range(n + 1)]
    visited[1] = True

    while queue:
        node, distance = queue.popleft()
        distances[node] = distance

        for next_node in graph[node]:
            if not visited[next_node]:
                queue.append((next_node, distance + 1))
                visited[next_node] = True

    return distances.count(max(distances))
